Source/InputMethodController.swift

```diff
@@ -0,0 +1,547 @@
+
+        let inputMode = keyHandler.inputMode
+        let optionKeyPressed = NSEvent.modifierFlags.contains(.option)
+
+        if inputMode == .bopomofo && optionKeyPressed {
+            let phaseReplacementItem = menu.addItem(withTitle: NSLocalizedString("Use Phrase Replacement", comment: ""), action: #selector(togglePhraseReplacement(_:)), keyEquivalent: "")
+            phaseReplacementItem.state = Preferences.phraseReplacementEnabled.state
+        }
+
+        menu.addItem(NSMenuItem.separator())
+        menu.addItem(withTitle: NSLocalizedString("User Phrases", comment: ""), action: nil, keyEquivalent: "")
+
+        if inputMode == .plainBopomofo {
+            menu.addItem(withTitle: NSLocalizedString("Edit Excluded Phrases", comment: ""), action: #selector(openExcludedPhrasesPlainBopomofo(_:)), keyEquivalent: "")
+        } else {
+            menu.addItem(withTitle: NSLocalizedString("Edit User Phrases", comment: ""), action: #selector(openUserPhrases(_:)), keyEquivalent: "")
+            menu.addItem(withTitle: NSLocalizedString("Edit Excluded Phrases", comment: ""), action: #selector(openExcludedPhrasesMcBopomofo(_:)), keyEquivalent: "")
+            if optionKeyPressed {
+                menu.addItem(withTitle: NSLocalizedString("Edit Phrase Replacement Table", comment: ""), action: #selector(openPhraseReplacementMcBopomofo(_:)), keyEquivalent: "")
+            }
+        }
+
+        menu.addItem(withTitle: NSLocalizedString("Reload User Phrases", comment: ""), action: #selector(reloadUserPhrases(_:)), keyEquivalent: "")
+        menu.addItem(NSMenuItem.separator())
+
+        menu.addItem(withTitle: NSLocalizedString("Check for Updates…", comment: ""), action: #selector(checkForUpdate(_:)), keyEquivalent: "")
+        menu.addItem(withTitle: NSLocalizedString("About McBopomofo…", comment: ""), action: #selector(showAbout(_:)), keyEquivalent: "")
+        return menu
+    }
+
+    // MARK: - IMKStateSetting protocol methods
+
+    override func activateServer(_ client: Any!) {
+        UserDefaults.standard.synchronize()
+
+        // Override the keyboard layout. Use US if not set.
+        (client as? IMKTextInput)?.overrideKeyboard(withKeyboardNamed: Preferences.basisKeyboardLayout)
+        // reset the state
+        currentDeferredClient = nil
+        currentCandidateClient = nil
+
+        keyHandler.clear()
+        keyHandler.syncWithPreferences()
+        self.handle(state: .Empty(), client: client)
+        (NSApp.delegate as? AppDelegate)?.checkForUpdate()
+    }
+
+    override func deactivateServer(_ client: Any!) {
+        keyHandler.clear()
+        self.handle(state: .Empty(), client: client)
+        self.handle(state: .Deactivated(), client: client)
+    }
+
+    override func setValue(_ value: Any!, forTag tag: Int, client: Any!) {
+        let newInputMode = InputMode(rawValue: value as? String ?? InputMode.bopomofo.rawValue)
+        LanguageModelManager.loadDataModel(newInputMode)
```

Comment:
Could you check if `setValue` is called only once per instantiation of the input method controller? Or could this be called every time we activate/deactivate the controller?

Since `LanguageModelManager.loadDataModel()` doesn't check if the data model is already loaded, I just wanted to make sure that we are not re-loading data models repeatedly.

Reply:
OK. I did additional checks to prevent loading data models repeatedly.